### PR TITLE
Proof of concept: Protobuf

### PR DESCRIPTION
**Pull request by huwpascoe**
_Protocol buffers are a flexible, efficient, automated mechanism for serializing structured data – think XML, but smaller, faster, and simpler. You define how you want your data to be structured once, then you can use special generated source code to easily write and read your structured data to and from a variety of data streams and using a variety of languages. You can even update your data structure without breaking deployed programs that are compiled against the "old" format._
 -- https://developers.google.com/protocol-buffers/docs/overview

Any object can be serialized, you no longer have to rely on cramming information into limited fields WINAPI-style. Instead, order classes can be tailored to suit the trait. Because object identification is based on a lookup table provided by the ObjectCreator, this allows huge bandwidth savings compared to the legacy string based identification tags. A typical move command in the current state might serialize as `MoveAAAABXXXXYYYY` but a ProtoContract version could generally be this sort of length: `SMAABXY`, less than half the size, easily saving megabytes of bandwidth.

Here's an example of the legacy order format as a data contract:

```
    [ProtoContract]
    public struct CPos : IScriptBindable, ILuaAdditionBinding, ILuaSubtractionBinding, ILuaEqualityBinding, ILuaTableBinding, IEquatable<CPos>
    {
        [ProtoMember(1)] public readonly int X;
        [ProtoMember(2)] public readonly int Y;
...
    }

    [ProtoContract]
    sealed class LegacyOrder
    {
        [ProtoMember(1)] public string OrderString;
        [ProtoMember(2)] public uint SubjectID;
        [ProtoMember(3)] public bool Queued;
        [ProtoMember(4)] public uint TargetActorID;
        [ProtoMember(5)] public CPos TargetLocation;
        [ProtoMember(6)] public string TargetString;
        [ProtoMember(7)] public CPos ExtraLocation;
        [ProtoMember(8)] public uint ExtraData;
    }
```

Contracts can be used to upgrade the OpenRA packet format, enforcing compile-time safety and even compressing the top level fields such as that "Frame" integer.

```
    [ProtoContract]
    sealed class Packet
    {
        [ProtoMember(1)] public readonly uint ClientID;
        [ProtoMember(2)] public readonly uint Frame;
        [ProtoMember(3)] public readonly byte[] Data;
    }
```

### CODE REVIEW

**Comment by Mailaender** [comment]
Interesting, can you add the license and developer of the new thirdparty library to the AUTHORS file?

File: OpenRA.Game/Network/Order.cs [hunk 1]
```diff
@@ -12,26 +12,21 @@
...
 using System.Linq;
 using OpenRA.Network;
+using ProtoBuf;
 
 namespace OpenRA
 {
-	[Flags]
-	enum OrderFields : byte
+	[ProtoContract]
+	sealed class LegacyOrder
```
Comment (pchote) [inline OpenRA.Game/Network/Order.cs]: It would be better to keep this named `Order` until/unless somebody actively starts migrating away from these.

**Comment by pchote** [comment]
This does look interesting.  It would be useful to have a couple of big test games before merging this, so that we can get a feeling for the bandwidth improvement and verify the cross-platform compatibility.

**Comment by huwpascoe** [comment]
Well it's proof of concept ATM, I'm gonna need to change a few more things to get it merge-ready.

**Comment by huwpascoe** [comment]
- The issue with the tests currently is how the ObjectCreator is referenced statically in OpenRA.Game.ModData.ObjectCreator, meaning it only works when run from OpenRA.Game. Need to find a way to fix that.

- There's not much point merging in the current state if it's just wrapping what's already there, save a few bytes from the varint fields. As the PR stands it's still sending the OrderString as a field opposed to utilizing the TypeResolver on the Contract's assembly name. Need a clear path towards the tailored orders while supporting the legacy.

**Comment by huwpascoe** [comment]
- And also the IsImmediate - a misnomer for a client-received ServerOrder - needs to be handled more effectively too.

**Comment by LipkeGu** [comment]
This seems to be increase Performance dramasticaly ? i have seen this style a few days ago in Java on Terasology (a minecraft clone) and they said they got an huge Performance increase :)

**Comment by chrisforbes** [comment]
@LipkeGu needs to be tested in context -- that other game may have been doing it a really bad way, etc. There's certainly scope for big wins here though.

I may not be able to eyeball the raw order stream any more though, which would be a shame ;)

**Comment by pchote** [comment]
@huwpascoe: This looks like a good architectural move to me, so I don't have any further comments until you're ready to get this merged.

**Comment by huwpascoe** [comment]
@chrisforbes If you extract the .proto definitions you can easily interpret the data with a python script or something : )

**Comment by pchote** [comment]
There's not much else we can do here for now, so closing.

**Pull request closed without merging**